### PR TITLE
Simplified AlertIOS

### PR DESCRIPTION
**Pull request by corbt**
Ok, so this started as fixing #5273 but ended up getting a little more complicated. :smile: 

Currently, AlertIOS has the following API:

* `alert(title, message, buttons, type)`
* `prompt(title, defaultValue, buttons, callback)`

I've changed the API to look like the following:

* `alert(title, message, callbackOrButtons)`
* `prompt(title, message, callbackOrButtons, type, defaultValue)`

I know that breaking changes are a big deal, but I find the current alert API to be fairly inconsistent and unnecessarily confusing. I'll try to justify my changes one by one:

1. Currently `type` is an optional parameter of `alert`. However, the only reason to change the alert type from the default is in order to create one of the input dialogs (text, password or username/password). So we're in a weird state where if you want a normal text input, you use `prompt`, but if you want a password input you use `alert` with the 'secure-text' type. I've moved `type` to `prompt` so all text input is now done with `prompt`.
2. Currently the only way to fire a callback when a user dismisses an `alert` is by passing in custom buttons. However, `prompt` allows you to pass in a callback that is called when the user clicks the "OK" button, without having to define custom buttons. But, you have to pass `null` to either the callback or buttons parameter (can't define both). Since these are mutually exclusive, I've combined them into one parameter and added analogous functionality to both `alert` and the `prompt` to make their APIs more consistent.
3. The defaultValue functionality of `prompt` got broken between 0.16 and 0.17. I've added it back in as an explicit argument (the default value reused the `message` field before, but there are times when you might want both a default value and a message, so they should be separate).

Those were all the major changes I made. I did also refactor the UIExplorer a bit to make it more clear what's going on in each example, and updated the documentation for AlertIOS.

### CODE REVIEW

**Comment by satya164** [comment]
I agree that current Alert API is confusing. But this is a very big breaking change, since alert is supposedly used in lots of apps. If we want to introduce a breaking change, we should provide a codemod to migrate old APIs to new one.

Also we should change the Android API as well.

Thoughts @ide @brentvatne @mkonicek ?

**Comment by corbt** [comment]
The only difference between this and the cross-platform Alert API is support for a callback in place of buttons. I can add that to the Android implementation though, I just figured this PR was getting big enough on its own. :neutral_face: 

And actually the "breaking" part of the change is pretty minor. The most common alert call (`alert(title, message, [buttons])`) doesn't change at all. Also, moving the position of the prompt's `defaultValue` shouldn't break many peoples' apps, because `defaultValue` has been broken since 0.17 and no one has complained about it. :smile: 

Only the following usages will need to be migrated with this change:
1. `prompt(title, message, null, callback)` -> `prompt(title, message, callback)`. Note that this doesn't make the API less configurable, just less confusing, because you weren't allowed to define both buttons and a callback anyway.
2. `alert(title, message, buttons, type)` -> `prompt(title, message, buttons, type)`. This change may actually affect people in the wild who were using password/login input types. But as I explain above, the only reason to change the type is if you want to use one of the text input alerts. If we're going to have a separate `prompt` function anyway (which I think is a clearer API), these types should live on `prompt`.

Is there an example of a breaking change with a codemod I could examine to see how to write/integrate that? I could also shim support for the old API for one version with deprecation warnings, which people may be more familiar with.

**Comment by satya164** [comment]
@corbt I've not seen many breaking changes, so no idea about it. I'm also working on #5336 , which will introduce a new API, and we need to decide a proper way of deprecating old APIs. I'm not sure what's the best way to do it though :(

**Comment by corbt** [comment]
Ok I've updated this to add support for the old API with `console.warn` deprecation notices, so unless I've missed something this shouldn't break anyone's apps. @mkonicek anyone at FB able to take a look at this? May affect your internal apps.

File: Libraries/Utilities/Alert.js [hunk 1]
```diff
@@ -68,10 +67,9 @@ class Alert {
     title: ?string,
     message?: ?string,
     buttons?: Buttons,
-    type?: AlertType
```
Comment (mkonicek) [inline Libraries/Utilities/Alert.js]: This was for convenience on iOS. How can people pass the type / style with this change?

File: Libraries/Utilities/AlertIOS.js [hunk 2]
```diff
@@ -27,61 +26,160 @@ export type AlertButtonStyle = $Enum<{
...
 /**
- * Launches an alert dialog with the specified title and message.
- *
- * Optionally provide a list of buttons. Tapping any button will fire the
- * respective onPress callback and dismiss the alert. By default, the only
- * button will be an 'OK' button.
+ * The AlertsIOS utility provides two functions: `alert` and `prompt`. All
+ * functionality available through `AlertIOS.alert` is also available in the
+ * cross-platform `Alert.alert`, which we recommend you use if you don't need
+ * iOS-specific functionality.
```
Comment (mkonicek) [inline Libraries/Utilities/AlertIOS.js]: >  All functionality available through `AlertIOS.alert` is also available in the cross-platform `Alert.alert`

Is this true for the `prompt` function?

**Comment by mkonicek** [comment]
Thanks for working on it! Yup a quick search tells me we have about a dozen files at fb that use `AlertIOS`.

File: Libraries/Utilities/AlertIOS.js (hunk 2 above)
Comment (corbt) [inline Libraries/Utilities/AlertIOS.js]: No, there isn't currently a cross-platform `Alert.prompt`, although there's no reason why this has to be iOS-specific.

File: Libraries/Utilities/Alert.js (hunk 1 above)
Comment (corbt) [inline Libraries/Utilities/Alert.js]: The only reason to change the `type` is to switch from an alert to a prompt, and I think it's very unlikely that anyone was using a dialog that appears as a prompt on one platform and an alert on another. But I can add that arg back in with a deprecation warning.

**Comment by corbt** [comment]
@mkonicek I've restored support for a `type` parameter in Alert.alert() with a deprecation warning.

**Comment by mkonicek** [comment]
Thanks for cleaning up the API! Like the new one much better.

This is now 100% backwards compatible, correct?

**Comment by corbt** [comment]
@mkonicek yes, with deprecation warnings.

**Comment by satya164** [comment]
Idea: In places where we've deprecation warnings, we can add a comment `// TODO: deprecated`, which will help us find these and remove entirely later.

**Comment by mkonicek** [comment]
Nice! If you guys are confident this definitely doesn't break any possible existing call site feel free to comment `@facebook-github-bot shipit`. Hope I'm not going to have a hard time being pinged by product teams at fb telling me their alerts stopped working :)

File: Libraries/Utilities/AlertIOS.js [hunk 3]
```diff
@@ -27,61 +26,160 @@ export type AlertButtonStyle = $Enum<{
...
+   *  - **callbackOrButtons** -- This optional argument should be either a
+   *    single-argument function or an array of buttons. If passed a function,
+   *    it will be called with the prompt's value when the user taps 'OK'.
+   *
+   *    If passed an array of button configurations, each button should include
+   *    a `text` key, as well as optional `onPress` and `style` keys (see example).
+   *    `style` should be one of 'default', 'cancel' or 'destructive'.
+   *  - **type**: string -- This configures the text input. One of 'plain-text',
+   *    'secure-text' or 'login-password'.
+   *  - **defaultValue**: string -- the default value for the text field.
```
Comment (mkonicek) [inline Libraries/Utilities/AlertIOS.js]: nit: I like the emphasis for the names but not sure it's a convention in the docs. Check other files. Would rather keep it consistent with docs for other files and then migrate all at once.

**Comment by corbt** [comment]
@mkonicek so as near as I can tell there isn't a great story for documenting function arguments in the RN docs right now. The only API I could find that had anything more formal than a prose definition is [LayoutAnimation](http://facebook.github.io/react-native/docs/layoutanimation.html#content), which uses the `@param` annotation. I've updated this PR to use the same annotation. Long-term it would be good to add formatting to the docs generator to deal with `@param` or an equivalent keyword nicely.

**Comment by mkonicek** [comment]
Oh but `@param` is not going to render well on the website now. Can you make this just a list please?

**Comment by corbt** [comment]
@mkonicek this better?

**Comment by mkonicek** [comment]
Sounds good, ship it :)

**Comment by mkonicek** [comment]
Yup that should render fine :) Looks good, ship it :+1:

**Comment by corbt** [comment]
@facebook-github-bot shipit

**Comment by Noitidart** [comment]
Is there a hidden `AlertAndroid.prompt`? I'm looking for the native prompt functionality on android. I looked inside `UIManager` but could not find a prompt for Android.

**Comment by ansarikhurshid786** [comment]
@Noitidart I am also looking for the same. have you got solution?

**Pull request closed without merging**